File: fast-tax-nodejs/REST/ft_response.js
/**
 * Input parameters for the GetBestMatch API call.
 */
export class GetBestMatchInput {
    constructor(data = {}) {
        this.Address = data.Address;
        this.Address2 = data.Address2;
        this.City = data.City;
        this.State = data.State;
        this.Zip = data.Zip;
        this.TaxType = data.TaxType;
        this.LicenseKey = data.LicenseKey;
        this.IsLive = data.IsLive !== undefined ? data.IsLive : true;
        this.TimeoutSeconds = data.TimeoutSeconds !== undefined ? data.TimeoutSeconds : 15;
    }

    toString() {
        return `GetBestMatchInput: Address = ${this.Address}, Address2 = ${this.Address2}, City = ${this.City}, State = ${this.State}, Zip = ${this.Zip}, TaxType = ${this.TaxType}, LicenseKey = ${this.LicenseKey}, IsLive = ${this.IsLive}, TimeoutSeconds = ${this.TimeoutSeconds}`;
    }
}

/**
 * Information Component for API responses.
 */
export class InformationComponent {
    constructor(data = {}) {
        this.Name = data.Name;
        this.Value = data.Value;
    }

    toString() {
        return `Name = ${this.Name}, Value = ${this.Value}`;
    }
}

/**
 * Error object for API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Desc = data.Desc;
        this.Number = data.Number;
        this.Location = data.Location;
    }

    toString() {
        return `Error: Desc = ${this.Desc}, Number = ${this.Number}, Location = ${this.Location}`;
    }
}

/**
 * Tax information for a matched address or ZIP code.
 */
export class BestMatchTaxInfo {
    constructor(data = {}) {
        this.Zip = data.Zip;
        this.City = data.City;
        this.County = data.County;
        this.StateAbbreviation = data.StateAbbreviation;
        this.StateName = data.StateName;
        this.TaxRate = data.TaxRate;
        this.StateRate = data.StateRate;
        this.CityRate = data.CityRate;
        this.CountyRate = data.CountyRate;
        this.CountyDistrictRate = data.CountyDistrictRate;
        this.CityDistrictRate = data.CityDistrictRate;
        this.SpecialDistrictRate = data.SpecialDistrictRate;
        this.InformationComponents = (data.InformationComponents || []).map(component => new InformationComponent(component));
        this.TotalTaxExempt = data.TotalTaxExempt;
        this.NotesCodes = data.NotesCodes;
        this.NotesDesc = data.NotesDesc;
    }

    toString() {
        const componentsString = this.InformationComponents.length
            ? this.InformationComponents.map(component => component.toString()).join(', ')
            : 'null';
        return `BestMatchTaxInfo: Zip = ${this.Zip}, City = ${this.City}, County = ${this.County}, StateAbbreviation = ${this.StateAbbreviation}, StateName = ${this.StateName}, TaxRate = ${this.TaxRate}, StateRate = ${this.StateRate}, CityRate = ${this.CityRate}, CountyRate = ${this.CountyRate}, CountyDistrictRate = ${this.CountyDistrictRate}, CityDistrictRate = ${this.CityDistrictRate}, SpecialDistrictRate = ${this.SpecialDistrictRate}, InformationComponents = [${componentsString}], TotalTaxExempt = ${this.TotalTaxExempt}, NotesCodes = ${this.NotesCodes}, NotesDesc = ${this.NotesDesc}`;
    }
}

/**
 * Response from GetBestMatch API, containing tax rate information for the best match.
 */
export class GetBestMatchResponse {
    constructor(data = {}) {
        this.TaxInfoItems = (data.TaxInfoItems || []).map(taxInfo => new BestMatchTaxInfo(taxInfo));
        this.MatchLevel = data.MatchLevel;
        this.Error = data.Error ? new Error(data.Error) : null;
        this.Debug = data.Debug || [];
    }

    toString() {
        const taxInfoItemsString = this.TaxInfoItems.length
            ? this.TaxInfoItems.map(taxInfo => taxInfo.toString()).join('; ')
            : 'null';
        const debugString = this.Debug.length
            ? this.Debug.join(', ')
            : 'null';
        return `GetBestMatchResponse: TaxInfoItems = [${taxInfoItemsString}], MatchLevel = ${this.MatchLevel}, Error = ${this.Error ? this.Error.toString() : 'null'}, Debug = [${debugString}]`;
    }
}

export default GetBestMatchResponse;

File: fast-tax-python/REST/ft_response.py
from dataclasses import dataclass
from typing import Optional, List


@dataclass
class GetBestMatchInput:
    Address: Optional[str] = None
    Address2: Optional[str] = None
    City: Optional[str] = None
    State: Optional[str] = None
    Zip: Optional[str] = None
    TaxType: Optional[str] = None
    LicenseKey: Optional[str] = None
    IsLive: bool = True
    TimeoutSeconds: int = 15

    def __str__(self) -> str:
        return (f"GetBestMatchInput: Address={self.Address}, Address2={self.Address2}, City={self.City}, "
                f"State={self.State}, Zip={self.Zip}, TaxType={self.TaxType}, LicenseKey={self.LicenseKey}, "
                f"IsLive={self.IsLive}, TimeoutSeconds={self.TimeoutSeconds}")


@dataclass
class InformationComponent:
    Name: Optional[str] = None
    Value: Optional[str] = None

    def __str__(self) -> str:
        return f"InformationComponent: Name={self.Name}, Value={self.Value}"


@dataclass
class Error:
    Desc: Optional[str] = None
    Number: Optional[str] = None
    Location: Optional[str] = None

    def __str__(self) -> str:
        return f"Error: Desc={self.Desc}, Number={self.Number}, Location={self.Location}"


@dataclass
class BestMatchTaxInfo:
    Zip: Optional[str] = None
    City: Optional[str] = None
    County: Optional[str] = None
    StateAbbreviation: Optional[str] = None
    StateName: Optional[str] = None
    TaxRate: Optional[float] = None
    StateRate: Optional[float] = None
    CityRate: Optional[float] = None
    CountyRate: Optional[float] = None
    CountyDistrictRate: Optional[float] = None
    CityDistrictRate: Optional[float] = None
    SpecialDistrictRate: Optional[float] = None
    InformationComponents: Optional[List['InformationComponent']] = None
    TotalTaxExempt: Optional[str] = None
    NotesCodes: Optional[str] = None
    NotesDesc: Optional[str] = None

    def __post_init__(self):
        if self.InformationComponents is None:
            self.InformationComponents = []

    def __str__(self) -> str:
        components_string = ', '.join(str(component) for component in self.InformationComponents) if self.InformationComponents else 'None'
        return (f"BestMatchTaxInfo: Zip={self.Zip}, City={self.City}, County={self.County}, "
                f"StateAbbreviation={self.StateAbbreviation}, StateName={self.StateName}, "
                f"TaxRate={self.TaxRate}, StateRate={self.StateRate}, CityRate={self.CityRate}, "
                f"CountyRate={self.CountyRate}, CountyDistrictRate={self.CountyDistrictRate}, "
                f"CityDistrictRate={self.CityDistrictRate}, SpecialDistrictRate={self.SpecialDistrictRate}, "
                f"InformationComponents=[{components_string}], TotalTaxExempt={self.TotalTaxExempt}, "
                f"NotesCodes={self.NotesCodes}, NotesDesc={self.NotesDesc}")


@dataclass
class GetBestMatchResponse:
    TaxInfoItems: Optional[List['BestMatchTaxInfo']] = None
    MatchLevel: Optional[str] = None
    Error: Optional['Error'] = None
    Debug: Optional[List[str]] = None

    def __post_init__(self):
        if self.TaxInfoItems is None:
            self.TaxInfoItems = []
        if self.Debug is None:
            self.Debug = []

    def __str__(self) -> str:
        tax_info_string = '; '.join(str(tax_info) for tax_info in self.TaxInfoItems) if self.TaxInfoItems else 'None'
        debug_string = ', '.join(self.Debug) if self.Debug else 'None'
        error = str(self.Error) if self.Error else 'None'
        return (f"GetBestMatchResponse: TaxInfoItems=[{tax_info_string}], MatchLevel={self.MatchLevel}, "
                f"Error={error}, Debug=[{debug_string}]")
